realm/src/main/java/io/realm/RealmObject.java

```diff
@@ -124,4 +135,142 @@ protected static Realm getRealm(RealmObject obj) {
     protected static Row getRow(RealmObject obj) {
         return obj.row;
     }
+
+    /**
+     * Set the Future instance returned by the worker thread, we need this instance
+     * to force {@link #load()} an async query, we use it to determine if the current
+     * RealmResults is a sync or async one.
+     *
+     * @param pendingQuery pending query.
+     */
+    void setPendingQuery(Future<Long> pendingQuery) {
+        this.pendingQuery = pendingQuery;
+        if (isLoaded()) {
+            // the query completed before RealmQuery
+            // had a chance to call setPendingQuery to register the pendingQuery (used btw
+            // to determine isLoaded behaviour)
+            onCompleted();
+
+        } // else, it will be handled by the Realm#handler
+    }
+
+    /**
+     * Determine if the current RealmObject is obtained synchronously or asynchronously (from
+     * a worker thread). Synchronous RealmObjects are by definition blocking hence this method
+     * will always return {@code true} for them.
+     * Note: This will return {@code true} if called from a standalone object (created outside of Realm).
+     *
+     * @return {@code true} if the query has completed and the data is available {@code false} if the
+     * query is in progress.
+     */
+    public boolean isLoaded() {
+        if (realm == null) {
+            return true;
+        }
+        realm.checkIfValid();
+        return pendingQuery == null || isCompleted;
+    }
+
+    /**
+     * Make an asynchronous query blocking.
+     * Note: This will return {@code true} if called from a standalone object (created outside of Realm).
+     *
+     * @return {@code true} if it successfully completed the query, {@code false} otherwise.
+     */
+    public boolean load() {
+        if (realm == null) {
+            return true;
+        }
+        // doesn't guarantee to import correctly the result (because the user may have advanced)
+        // in this case the Realm#handler will be responsible for retrying
+        realm.checkIfValid();
+        if (pendingQuery != null && !pendingQuery.isDone() && !isCompleted) {
+            return onCompleted();
+        }
+        return false;
+    }
+
+    /**
+     * Called to import the handover row pointer & notify listeners.
+     *
+     * @return {@code true} if it successfully completed the query, {@code false} otherwise.
+     */
+    boolean onCompleted() {
+        realm.checkIfValid();
```

Comment:
Not a big problem, but `realm.checkIfValid()` might be called twice? It will be called in `void onCompleted(Long handoverRowPointer)` so we can remove the checking here?